skimage/registration/_lddmm.py

```diff
@@ -0,0 +1,2211 @@
+"""
+This is an implementation of the LDDMM algorithm with modifications, written
+by Devin Crowley and based on "Diffeomorphic registration with intensity
+transformation and missing data: Application to 3D digital pathology of
+Alzheimer's disease."
+
+This paper extends on an older LDDMM paper, "Computing large deformation
+metric mappings via geodesic flows of diffeomorphisms."
+
+This is the more recent paper:
+Tward, Daniel, et al. "Diffeomorphic registration with intensity
+transformation and missing data: Application to 3D digital pathology of
+Alzheimer's disease." Frontiers in neuroscience 14 (2020).
+https://doi.org/10.3389/fnins.2020.00052
+
+This is the original LDDMM paper:
+Beg, M. Faisal, et al. "Computing large deformation metric mappings via
+geodesic flows of diffeomorphisms." International journal of computer vision
+61.2 (2005): 139-157.
+https://doi.org/10.1023/B:VISI.0000043755.93987.aa
+"""
+
+import warnings
+import numpy as np
+from collections import namedtuple
+
+from scipy.interpolate import interpn
+from scipy.linalg import inv, solve, det, svd
+from skimage.transform import resize, rescale
+
+from ._lddmm_utilities import _validate_scalar_to_multi
+from ._lddmm_utilities import _validate_ndarray
+from ._lddmm_utilities import _validate_spacing
+from ._lddmm_utilities import _compute_axes
+from ._lddmm_utilities import _compute_coords
+from ._lddmm_utilities import _multiply_coords_by_affine
+from ._lddmm_utilities import _compute_tail_determinant
+from ._lddmm_utilities import resample
+from ._lddmm_utilities import sinc_resample
+
+r"""
+  _            _       _
+ | |          | |     | |
+ | |        __| |   __| |  _ __ ___    _ __ ___
+ | |       / _` |  / _` | | '_ ` _ \  | '_ ` _ \
+ | |____  | (_| | | (_| | | | | | | | | | | | | |
+ |______|  \__,_|  \__,_| |_| |_| |_| |_| |_| |_|
+
+"""
+
+
+class _Lddmm:
+    """
+    Class for storing shared values and objects used in learning a
+    registration at a single scale.
+    """
+
+    def __init__(
+        self,
+        # Images.
+        reference_image,
+        moving_image,
+        # Image spacings.
+        reference_image_spacing=None,
+        moving_image_spacing=None,
+        # Iterations.
+        num_iterations=None,
+        num_affine_only_iterations=None,
+        num_rigid_affine_iterations=None,
+        # Stepsizes.
+        affine_stepsize=None,
```

Comment:
for the different parameters of the constructor, could you use the default values which you are setting below instead of `None` ? Then you could do 
`self.__dict__.update(vars())`

I'm trying to make the code a bit shorter :-).

Reply:
I had considered that, but I believe there are three issues with it - 
- some of the parameter defaults depend on the values of other parameters,
- some of the parameter defaults are big and constructed in multiple lines, &
- all of the validation happens as the inputs are set as attributes.

Reply:
However, the defaults for the parameters in the diffeomorphic_metric_mapping signature could be provided there instead of None.